network/p2p/gossip/handler.go

```diff
@@ -0,0 +1,85 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+
+	bloomfilter "github.com/holiman/bloomfilter/v2"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p"
+	"github.com/ava-labs/avalanchego/network/p2p/gossip/proto/pb"
+)
+
+var (
+	_ p2p.Handler = (*Handler[Gossipable])(nil)
+
+	ErrInvalidHash = errors.New("invalid hash")
+)
+
+func NewHandler[T Gossipable](set Set[T], maxResponseSize int) *Handler[T] {
+	return &Handler[T]{
+		Handler:         p2p.NoOpHandler{},
+		set:             set,
+		maxResponseSize: maxResponseSize,
+	}
+}
+
+type Handler[T Gossipable] struct {
+	p2p.Handler
+	set             Set[T]
+	maxResponseSize int
+}
+
+func (h Handler[T]) AppRequest(_ context.Context, _ ids.NodeID, _ time.Time, requestBytes []byte) ([]byte, error) {
+	request := &pb.PullGossipRequest{}
+	if err := proto.Unmarshal(requestBytes, request); err != nil {
+		return nil, err
+	}
+
+	if len(request.Salt) != hashLength {
+		return nil, ErrInvalidHash
+	}
+	filter := &BloomFilter{
+		Bloom: &bloomfilter.Filter{},
+		Salt:  Hash{},
+	}
+	if err := filter.Bloom.UnmarshalBinary(request.Filter); err != nil {
+		return nil, err
+	}
+
+	copy(filter.Salt[:], request.Salt)
+
+	// filter out what the requesting peer already knows about
+	unknown := h.set.Get(func(gossipable T) bool {
+		return !filter.Has(gossipable)
+	})
```

Comment:
Can we change this so that `.Get` handles the maximum message size? I don't really like that we get a bunch of data and then potentially drop it all here (rather than just stopping the collection of the data earlier)

Reply:
Yeah we'll have to unmarshal each item to make sure it's not exceeding the max size... so I think we'll have to change `Get` from returning stuff to `Iterate` and we can create `gossipBytes` in there.

Reply:
Alternatively - we just return the marshalled form from the set. I'm happy with either.